packages/swapper/src/swappers/PortalsSwapper/getPortalsTradeQuote/getPortalsTradeQuote.ts

```diff
@@ -0,0 +1,207 @@
+    const portalsTradeOrderResponse = await fetchPortalsTradeOrder({
+      sender: sendAddress,
+      inputToken,
+      outputToken,
+      inputAmount: sellAmountIncludingProtocolFeesCryptoBaseUnit,
+      slippageTolerancePercentage: Number(slippageTolerancePercentageDecimal) * 100,
+      partner: getTreasuryAddressFromChainId(sellAsset.chainId),
+      feePercentage: affiliateBpsPercentage,
+      validate: true,
+    }).catch(e => {
+      console.info('failed to get Portals quote with validation enabled', e)
+
+      // If validation fails, try again without validation, we won't get network fees, but we can't do any better
+      return fetchPortalsTradeOrder({
+        sender: sendAddress,
+        inputToken,
+        outputToken,
+        inputAmount: sellAmountIncludingProtocolFeesCryptoBaseUnit,
+        slippageTolerancePercentage: Number(slippageTolerancePercentageDecimal) * 100,
+        partner: getTreasuryAddressFromChainId(sellAsset.chainId),
+        feePercentage: affiliateBpsPercentage,
+        validate: false,
+      })
+    })
+
+    const {
+      context: {
+        orderId,
+        outputAmount: buyAmountAfterFeesCryptoBaseUnit,
+        minOutputAmount: buyAmountBeforeFeesCryptoBaseUnit,
+        slippageTolerancePercentage,
+        target: allowanceContract,
+        feeAmount,
+        gasLimit,
+      },
+    } = portalsTradeOrderResponse
+
+    const rate = bn(buyAmountAfterFeesCryptoBaseUnit)
+      .div(input.sellAmountIncludingProtocolFeesCryptoBaseUnit)
+      .toString()
+
+    const adapter = assertGetEvmChainAdapter(chainId)
+    const { average } = await adapter.getGasFeeData()
+
+    const networkFeeCryptoBaseUnit = calcNetworkFeeCryptoBaseUnit({
+      ...average,
+      supportsEIP1559,
+      // times 1 isn't a mistake, it's just so we can write this comment above to mention that Portals already add a
+      // buffer of ~15% to the gas limit
+      gasLimit: bnOrZero(gasLimit).times(1).toFixed(),
+    })
+
+    const tradeQuote: TradeQuote = {
+      id: orderId,
+      receiveAddress: input.receiveAddress,
+      affiliateBps,
+      potentialAffiliateBps,
+      rate,
+      slippageTolerancePercentageDecimal: (slippageTolerancePercentage / 100).toString(),
+      steps: [
+        {
+          accountNumber,
+          allowanceContract,
+          rate,
+          buyAsset,
+          sellAsset,
+          buyAmountBeforeFeesCryptoBaseUnit,
+          buyAmountAfterFeesCryptoBaseUnit,
+          sellAmountIncludingProtocolFeesCryptoBaseUnit:
+            input.sellAmountIncludingProtocolFeesCryptoBaseUnit,
+          feeData: {
+            networkFeeCryptoBaseUnit,
+            // Protocol fees are always denominated in buy asset here, this is the downside on the swap
+            protocolFees: {
+              [buyAsset.assetId]: {
+                amountCryptoBaseUnit: feeAmount,
+                asset: buyAsset,
+                requiresBalance: false,
```

Comment:
So the swap will take from the amountOut right?

Reply:
Precisely! That's the downside on the swap